xclib/evaluation/xc_metrics.py

```diff
@@ -650,3 +649,139 @@ def __init__(self, true_labels, inv_propensity_scores=None,
             "Metrices() is deprecated; use Metrics().",
             category=FutureWarning)
         super().__init__(true_labels, inv_propensity_scores, remove_invalid)
+
+@nb.njit(parallel=True)
+def _fast_recall_with_indices(pred_indices, true_indices, true_indptr, k):
+    m = pred_indices.shape[0]
+    cum_intersections = np.zeros((m, k), dtype=np.float64)
+    gt_cardinality = np.zeros((m, 1), dtype= np.float64)
+    for i in nb.prange(pred_indices.shape[0]):
+        intersection = in1d(pred_indices[i][:k], np.unique(true_indices[true_indptr[i]: true_indptr[i + 1]]))
+        cum_intersections[i] = np.cumsum(intersection)
+        gt_cardinality[i, 0] = true_indptr[i + 1] - true_indptr[i]
+
+    recall = mean_rows(np.multiply(cum_intersections, 1 / gt_cardinality))
+    micro_recall = np.sum(cum_intersections, axis = 0) / np.sum(gt_cardinality)
+    return recall, micro_recall
+
+def fast_recall_with_indices(X_indices, true_labels, k):
+    """
+    Compute recall@k, microR@k for values 1...k, faster than using `recall`
+    Arguments:
+    ----------
+    X_indices: np.ndarray
+        2D numpy array with indices sorted in descending order according to score for each row
+    true_labels: csr_matrix
+        ground truth in sparse format
+    k: int
+        compute psprecision for ints in [1, k]
+    Returns:
+    -------
+    np.ndarray: recall values
+    """
+    return _fast_recall_with_indices(X_indices, true_labels.indices.astype(np.int64), true_labels.indptr, k)
+
+@nb.njit(parallel=True)
+def restict_preds_for_gt_calc(pred_indices, num_gt, k):
+    """
+    Returns top min(GT, k) indices for each data point
+    Arguments:
+    ----------
+    X_indices: np.ndarray
+        2D numpy array with indices sorted in descending order according to score for each row
+    num_gt: np.array
+        number of ground truth for each element
+    k: int
+        max number of preds to consider
+    Returns:
+    -------
+    np.ndarray:  top min(GT, k) indices for each data point
+    """
+    num_docs = pred_indices.shape[0]
+    high_lim = 100000000
+    restricted_preds = pred_indices + high_lim # ensures that indices that are not overwritten will not intersect with any GT
+    for doc_indx in nb.prange(num_docs):
+        restrict_indx = min(k, num_gt[doc_indx])
+        restricted_preds[doc_indx][: restrict_indx] = pred_indices[doc_indx][:restrict_indx]
+    return restricted_preds
+
+def calc_gt_metrics(pred_indices, true_labels, k):
```

Comment:
User should be able to compute recall@gt and microR@gt separately. Moreover, what if user has predictions as 'csr_matrix' matrix or as 2 nd-arrays (indices and values)? Then old functions including precision, recall, and nDCG etc., would work but this won't. 

Support multiple formats (see https://github.com/kunaldahiya/pyxclib/blob/74da00fef66cddd12e40406f723df07eb4d9a120/xclib/evaluation/xc_metrics.py#L230)?

Reply:
Added 2 new functions `micro_recall_at_gt` and `recall_at_gt` along with their corresponding helper functions with `_` prefix, this now supports for `csr_matrix` and `np.2darray`, like other functions.